jurity/recommenders/doubly_robust.py

```diff
@@ -0,0 +1,150 @@
+
+from typing import List, Union, Tuple, Optional
+
+import numpy as np
+import pandas as pd
+
+from jurity.recommenders.base import _BaseRecommenders
+from jurity.utils import Constants, get_sorted_clicks
+
+
+class DoublyRobust(_BaseRecommenders):
+    """Doubly Robust Estimator
+
+    Calculates the DR, an estimate of CTR that combines the directly predicted values with a correction based on how
+    likely an item was to be recommended by the historic policy if the user saw the item in the historic data.
+
+    ..math::
+        DR = \frac{1}{n} \sigma (\hat{r_a} + \frac{(r_a -\hat{r_a}) I(\hat{a} = a}{p(a|x,h)})
+
+    In this equation:
+    * n is the total size of the test data
+    * r_a is the observed reward
+    * \hat{r_a} is the predicted reward
+    * hat{a} is the recommended item
+    * I(\hat{a} = a} is a boolean of whether the user-item pair has historic data
+    * p(a|x,h) is the probability of the item being recommended for the test context given the historic data
+
+    At a high level, doubly robust estimation combines a direct estimate with an IPS-like correction if historic data is
+    available. If historic data is not available, the second term is 0 and only the predicted reward is used for the
+    user-item pair.
+
+    """
+
+    def __init__(self, click_column: str, k: Optional[int] = None, user_id_column: str = Constants.user_id,
+                 item_id_column: str = Constants.item_id, value_column: Optional[str] = None,
+                 propensity_column: Optional[str] = None):
+        """Initializes the DoublyRobust object
+
+        Parameters
+        ---------
+        click_column: str
+            The column to use for scoring the arms.
+        k: Optional[int]
+            The number of recommendations per user. If not specified, all recommendations will be used.
+        user_id_column: str
+            The column name for the user ids. Defaults to Constants.user_id.
+        item_id_column: str
+            The column name for the item ids. Defaults to Constants.item_id.
+        value_column: Optional[str]
+            The column to calculate the DR on. If different from ``click_column``, the recommendations will be sorted
+            by ``click_column`` and the DR will be calculated on the matching rows, but on the ``value_column``. If not
+            specified, ``click_column`` will be used.
+        propensity_column: Optional [str]
+            The column with historic item probabilities in the actual results. Defaults to Constants.propensity_column.
+            If column is not provided, a simple random policy with equal likelihood for every item will be assumed.
+        """
+        pass
+
+    def get_score(self, actual_results: pd.DataFrame, predicted_results: pd.DataFrame, batch_accumulate: bool = False,
+                  return_extended_results: bool = False) -> Union[float, dict, Tuple[float, float], Tuple[dict, dict]]:
+        """Evaluates the current metric on the given data.
```

Comment:
Not related to this PR, but the fact that we need to copy all this for each method doesn't seem ideal. @dorukkilitcioglu, I'm sure you don't like it?

Reply:
Copy what? Do you mean pydoc?

Reply:
@skadio the docstring for get_score is really long, and the same for every class even though the implementations are unique

Reply:
Oh I see.. In the java world, we used to do sth like "See @<other_class>.<main_method> " and then clicking on it would take you to javadoc of the main method in most IDEs. Maybe there is a Python equivalent of the same idea?

Reply:
I wonder if we can do more abstraction for the scoring class here? E.g., we keep this dosctring in baseRecommender class and docstring reads something like 

```
print(metric_of_interest.get_score(actual_responses_batch, recommendations_batch, return_extended_results=True))
            >>> {'metric_of_interest': 0.316, 'support': 122}
```
Then you could inject the actual scoring code in each child `get_score` as a closure, or an another private method _score_implementation() or something and get rid of the duplicate long docstrings?

Reply:
Yup, I do hate copying over the whole docstring :D I think @filip-michalsky 's suggestion makes sense, we can keep the docstring on the base class' `get_score`, add a new abstract method `_get_score`, and let `get_score` call `_get_score` in the base class. Plus, this base `get_score` function could also automatically call `_get_correction` before passing the dataset to `_get_score`.

As long as the docstring shows up correctly in the IDE and in the docs, I'm good.